src/azul/plugins/repository/canned/__init__.py

```diff
@@ -69,7 +73,107 @@
 log = logging.getLogger(__name__)
 
 
-class CannedSourceRef(SourceRef[SimpleSourceSpec, 'CannedSourceRef']):
+@attrs.frozen(kw_only=True)
+class CannedSourceSpec(SourceSpec):
+    url: str
+    ref: str
+    path: Sequence[str]
+    sep: ClassVar[str] = '::'
+
+    @classmethod
+    def parse(cls, spec: str) -> Self:
+        """
+        Construct an instance from its string representation, using the syntax
+        '{repo-url}::{ref}::{path}::{prefix}' where:
+        {repo-url} is the HTTPS URL of a GitHub repository's `.git` file. It
+                   must match the format https://github.com/USER/REPO.git
+        {ref} is a branch name, tag, or commit SHA
+        {path} is the path to the base of the staging area (e.g. 'tests/area1')
+        {prefix} is the common prefix and partition length (e.g. '/0')
+
+        >>> s = CannedSourceSpec.parse('https://github.com/user/repo.git::f0f0f0f0::dir/subdir::/0')
```

Comment:
```suggestion
        >>> s = CannedSourceSpec.parse('github.com/user/repo.git:f0f0f0f0:dir/subdir')
```

The double separator looks ugly. I assume you used it because of the conflicting colon after URL schemes. Just modify the parser to use `rsplit` with `maxsplit`.

I also assume that we don't use these sources with any prefix other than `/0` so we can hard-code that for now. Otherwise, just point me to the use case.

Reply:
Also, both path ref can be expressed as a GitHub URL. Assuming I am right about the prefix, why do we need a separator in the first place?